### PR TITLE
Fill optional AST properties when both estree and typescript parser plugin are enabled (Part 2)

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Ref https://github.com/babel/babel/issues/16679
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR includes commits from #17224.

As a follow-up to #17224, in this PR we make the `cloneIdentifier` a parser method and move `castNodeTo` to the base parser. We also applied tokens hotfix to TS-ESLint AST so that the test can move forward. Eventually we will remove those hotfix once the upstream issue is resolved.

This PR also ensures support of two stage 3 proposals: decorators and using.

After this PR gets merged, I will run the eslint-parser against our own parser test suites, hopefully we can find more edge cases.

### CODE REVIEW

**Review by nicolo-ribaudo — APPROVED** [review]
(next time, it'd be helpful to have an empty `=== THIS PR STARTS HERE ===` commit to separate the commits of one PR from the other 😛)

**Comment by JLHwung** [comment]
Rebased, the diff should be much easier to review now.

File: eslint/babel-eslint-parser/src/convert/convertTokens.cts [hunk 1]
```diff
@@ -107,7 +107,7 @@ function convertToken(
   newToken.range = [token.start, token.end];
 
   if (label === tl.name) {
-    if (token.value === "static") {
+    if (token.value.match(/^(?:let|static|yield)$/)) {
```
Comment (liuxingbaoyu) [inline eslint/babel-eslint-parser/src/convert/convertTokens.cts:110]: It seems like `===` would be faster?
[perf](https://perf.link/#eyJpZCI6ImQ2amVicjg3cmVxIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6InZhciBhID0gXCJhYWFhYVwiXG52YXIgYiA9IFwic3RhdGljXCJcbnZhciBjOyIsInRlc3RzIjpbeyJuYW1lIjoic3RyIiwiY29kZSI6ImxldCB4ID0gU3RyaW5nKGEpO1xuYyA9IHggPT09IFwibGV0XCIgfHwgeCA9PT0gXCJzdGF0aWNcIiB8fCB4ID09PSBcInlpZWxkXCI7XG54ID0gU3RyaW5nKGIpO1xuYyA9IHggPT09IFwibGV0XCIgfHwgeCA9PT0gXCJzdGF0aWNcIiB8fCB4ID09PSBcInlpZWxkXCI7IiwicnVucyI6WzkzMzAwMCwyMzMzMDAwLDIwMDQwMDAsMTIyODQwMDAsOTMzMDAwLDIwOTg5MDAwLDczMDUwMDAsMTUzMzQwMDAsOTMzMDAwLDExMjU5MDAwLDc3NTAwMCwxMDkxMzAwMCw2MDQwMDAsMzg0MzAwMCwzMzUwMDAsMTk0MjEwMDAsMjExOTAwMCwxMDg3MDAwLDkzMzAwMCwxMzMwNzAwMCwzMDIwMDAsNzg0NTAwMCw4OTQwMDAwLDE3NzQwMDAwLDE2ODQ2MDAwLDIxNzMwMDAsMTU4OTAwMDAsMTUwMzEwMDAsMTM5OTAwMCwyMzQzMjAwMCwxNDY3MjAwMCwyMTAzMjAwMCwyMDc5MzAwMCwxMDQwMDAwLDEyMTUwMDAwLDIwMDAsMTI1MzgwMDAsNDM5NzAwMCwyMDg0MDAwLDExMzI1MDAwLDcxODYwMDAsNDI0MDAwLDcwODYwMDAsNTExNTAwMCwxNTI3MDAwLDE2MzI3MDAwLDIyMzUwMDAwLDIzNzQ0MDAwLDEzOTY0MDAwLDk4OTMwMDAsOTMzMDAwLDE0NzIwMDAsOTMzMDAwLDIwMDY3MDAwLDE3NTY1MDAwLDc1MzQwMDAsOTMzMDAwLDI1OTMzMDAwLDM2NTcwMDAsMTE4NDgwMDAsMTI0MzAwMCwyOTgwMDAsMTc0NjgwMDAsODk0NDAwMCw5MzMwMDAsMTQ5OTAwMDAsMTM0NjYwMDAsMTc4MjEwMDAsODYyMTAwMCwxMzIyMDAwMCwxOTc5MDAwMCwzNzU2MDAwLDE3NTk3MDAwLDEyNTI2MDAwLDE5MTI4MDAwLDk3MjAwMDAsNDE2NjAwMCw2MjM3MDAwLDEzOTI1MDAwLDUwNzgwMDAsNTY4NzAwMCwyMTU1MDAwMCw1Mzg2MDAwLDQyNDkwMDAsMzQxNTAwMCw3MTE0MDAwLDYwNzYwMDAsNjczMDAwLDYxNDAwMDAsNjc1ODAwMCwyMTc2MDAwMCwxMTE4MjAwMCwyMzUzNDAwMCwxOTM2MDAwMCwxOTEwODAwMCw0NjAzMDAwLDkwOTMwMDAsODEzODAwMCwxODI1MzAwMCwxMzEzMzAwMF0sIm9wcyI6OTU5OTA1MH0seyJuYW1lIjoicmVnZXgiLCJjb2RlIjoiYyA9IFN0cmluZyhhKS5tYXRjaCgvXig%2FOmxldHxzdGF0aWN8eWllbGQpJC8pXG5jID0gU3RyaW5nKGIpLm1hdGNoKC9eKD86bGV0fHN0YXRpY3x5aWVsZCkkLylcbiIsInJ1bnMiOlsxMzAwMDAsNzk1MDAwLDEwODUwMDAsOTUzMDAwLDEwMDAsNjAyODAwMCwxODMwMDAwLDQ1MTMwMDAsMjU3MDAwLDM2MzcwMDAsMzM0MDAwMCwyODIzMDAwLDEwMDAsMTQxMTAwMCwxMDAwLDM4NjUwMDAsNzk3MDAwLDU1MDAwMCw2NzAwMDAwLDM3MjgwMDAsMTAwMCwyMDMxMDAwLDI3NzMwMDAsNTgyNDAwMCwzOTMzMDAwLDEwMDAsNTA1MDAwMCwyNTAwMDAwLDU2MTAwMCw2NDYzMDAwLDQ4NTAwMDAsNjQ5MzAwMCw3MDc0MDAwLDE2MzAwMCwzNzQ4MDAwLDEwMDAsNDI0MDAwMCwxNDIzMDAwLDI4ODAwMCwzNzI4MDAwLDk1MzAwMCw3NTM0MDAwLDEwNjEwMDAsOTUzMDAwLDI5NjAwMCw5OTcwMDAsNzUyMTAwMCw2NDA2MDAwLDIzMDkwMDAsMzI1NDAwMCwxMzAwMCw1MzUwMDAsMTYwMDAsNjIwMTAwMCw0Njg0MDAwLDIzMDAwMDAsNzAwMCw2NzMyMDAwLDEyNjcwMDAsMzYzODAwMCwxMDAwLDczOTAwMDAsMzM1MTAwMCwxNDg4MDAwLDEwMDAsMzk0NDAwMCwyNzI1MDAwLDQyMzMwMDAsMTgxNjAwMCwzOTExMDAwLDQ1OTgwMDAsOTUzMDAwLDM2OTQwMDAsMzIxOTAwMCw2NTM5MDAwLDI3MzkwMDAsMTMwNzAwMCwyMjM0MDAwLDQyNzQwMDAsMTU3ODAwMCwxNzAxMDAwLDYyNTQwMDAsMTc0MzAwMCwxMjczMDAwLDEyMDEwMDAsMTg1MzAwMCwxMjczMDAwLDEwMDAsMTAxNDAwMCw5ODYwMDAsMTU3MzAwMCwyNzk1MDAwLDcwMzkwMDAsMjA1NjAwMCw2NjkzMDAwLDE0ODgwMDAsNDEwNDAwMCwyMzMzMDAwLDEwMDAsNDI0MDAwMF0sIm9wcyI6MjY5ODgzMH1dLCJ1cGRhdGVkIjoiMjAyNS0wNC0wOVQxNDozMDoyNS45ODRaIn0%3D)

**Review by liuxingbaoyu — APPROVED** [review]

File: eslint/babel-eslint-parser/src/convert/convertTokens.cts (hunk 1 above)
Comment (JLHwung) [reply eslint/babel-eslint-parser/src/convert/convertTokens.cts]: Thanks for the benchmark, revised.
